File: exercises/101-109-smallproblems/easy8/doublechar.rb

def repeater(str)
  doubled = ''
  str.chars.each { |char| doubled << char * 2 }
  doubled
end

def double_consonants(str)
  doubled = ''
  str.chars.each do |char|
    doubled << char
    doubled << char if consonant?(char)
  end
  doubled
end

def consonant?(char)
  char =~ /[a-zA-Z]/ && char =~ /[^aeiou]/
end

## Tests

p repeater('Hello') == "HHeelllloo"
p repeater("Good job!") == "GGoooodd  jjoobb!!"
p repeater('') == ''

p double_consonants('String') == "SSttrrinngg"
p double_consonants("Hello-World!") == "HHellllo-WWorrlldd!"
p double_consonants("July 4th") == "JJullyy 4tthh"
p double_consonants('') == ""

File: 210/small-problems/easy5/double-char.js
function repeater(string) {
  var doubled = '';

  string.split('')
        .forEach(function (char) {
           doubled += char + char;
        });

  return doubled;
}

function doubleConsonants(string) {
  // Based on Nick Johnson's regex:
  // return string.replace(/(?![aeiou0-9])\w/g, '$&$&');

  // Pete Hanson's suggestion:
  return string.replace(/(\w)/g, function(char) {
    return char.match(/[aeiou]/i)) ? char : (char + char);
  });
}

console.log(repeater('Hello') === 'HHeelllloo');
console.log(repeater('Good job!') === 'GGoooodd  jjoobb!!');
console.log(repeater('') === '');

console.log(doubleConsonants('String') === 'SSttrrinngg');
console.log(doubleConsonants('Hello-World!') === 'HHellllo-WWorrlldd!');
console.log(doubleConsonants('July 4th') === 'JJullyy 4tthh');
console.log(doubleConsonants('') === '');
